sea.js

```diff
@@ -1249,27 +1313,83 @@
       if(key === link_is(val)){ return eve.to.next(msg) } // and the ID must be EXACTLY equal to its property
       no("Alias not same!"); // that way nobody can tamper with the list of public keys.
     };
-    check.pub = function(eve, msg, val, key, soul, at, no, user, pub){ var tmp; // Example: {_:#~asdf, hello:'world'~fdsa}}
-      if('pub' === key && '~'+pub === soul){
+    check.pub = function(eve, msg, val, key, soul, at, no, user, pub){ var tmp // Example: {_:#~asdf, hello:'world'~fdsa}}
+      const raw = S.parse(val) || {}
+      
+      const verify = (certificate, certificant, cb) => {
+        if (certificate['m'] && certificate['s'] && certificant && pub) {
+          // now verify certificate
+          return SEA.verify(certificate, pub, data => { // check if "pub" (of the graph owner) really issued this cert
+            if (u !== data && u !== data.e && msg.put['>'] && msg.put['>'] > parseFloat(data.e)) return no("Certificate expired.")
+            // "data.c" = a list of certificants/certified users, "data.p" = a list of allowed patterns
+            if (u !== data && data.c && data.p && (data.c.indexOf('*') || data.c.indexOf(certificant))) {
+              // ok, now "certificant" is in the "certificants" list, but is "path" allowed? Check path
+              let path = soul + '/' + key
+              path = path.replace(path.substring(0, path.indexOf('/') + 1), '')
+              for (p of data.p) {
+                if (new RegExp(p).test(path)) {
+                  return cb(data)
+                }
+              }
+              return no("Certificate verification fail.")
+            }
+          })
+        }
+        return
+      }
+      
+      if ('pub' === key && '~'+pub === soul) {
         if(val === pub){ return eve.to.next(msg) } // the account MUST match `pub` property that equals the ID of the public key.
-        return no("Account not same!");
+        return no("Account not same!")
       }
-      if((tmp = user.is) && pub === tmp.pub){
-        SEA.sign(SEA.opt.pack(msg.put), (user._).sea, function(data){
+
+      if ((tmp = user.is) && tmp.pub && !raw['*'] && !raw['+']){
+        SEA.sign(SEA.opt.pack(msg.put), (user._).sea, function(data){ // needs to be refactored
           if(u === data){ return no(SEA.err || 'Signature fail.') }
-          if(tmp = link_is(val)){ (at.sea.own[tmp] = at.sea.own[tmp] || {})[pub] = 1 }
-          msg.put[':'] = JSON.stringify({':': tmp = SEA.opt.unpack(data.m), '~': data.s});
```

Comment:
This is changed to do without `JSON.stringify` call. Is this now 100% compatible with prior code?

Reply:
@mhelander I don't understand. JSON.stringify is still there at line 1355 and line 1367.